src/Serialization/Address.purs

```diff
@@ -0,0 +1,348 @@
+type PubKeyAddress = BaseAddress Ed25519KeyHash Ed25519KeyHash
+
+-- | Address where payment and stake delegation rights are bound to the same ScriptHash
+type ScriptAddress = BaseAddress ScriptHash ScriptHash
+
+
+-- | Helper for reclaiming `BaseAddress` type arguments during decoding.
+newtype CredType (p :: Type) = CredType (Maybe Boolean)
+
+----
+
+foreign import stakeCredFromKeyHash :: Ed25519KeyHash -> StakeCredentialCsl
+foreign import stakeCredFromScriptHash :: ScriptHash -> StakeCredentialCsl
+
+-- | Only use on csl types that supports CSL's to_address()
+foreign import toAddressCslUnsafe :: CslType -> AddressCsl
+
+foreign import newRewardAddressCsl
+  :: { network :: Int
+     , paymentStakeCred :: StakeCredentialCsl
+     } -> RewardAddressCsl
+
+foreign import newBaseAddressCsl
+  :: { network :: Int
+     , paymentStakeCred :: StakeCredentialCsl
+     , delegationStakeCred :: StakeCredentialCsl
+     } -> BaseAddressCsl
+
+foreign import addressBytesImpl :: AddressCsl -> Uint8Array
+foreign import addressBech32Impl :: AddressCsl -> Bech32String
+
+-- | Ensures that stake credentials are of requested type
+-- therefore allowing to set BaseAddress' type arguments
+foreign import headerCheckBaseAddr
+  :: forall p d
+   . MaybeFfiHelper
+  -> { payment :: CredType p
+     , delegation :: CredType d}
+  -> Uint8Array
+  -> BaseAddressCsl
+  -> Maybe (BaseAddress p d)
+
+-- | Ensures that stake credentials are of requested type
+-- therefore allowing to set RewardAddress' type arguments
+foreign import headerCheckRewardAddr
+  :: forall p
+   . MaybeFfiHelper
+  -> { payment :: CredType p}
+  -> Uint8Array
+  -> RewardAddressCsl
+  -> Maybe (RewardAddress p)
+
+foreign import baseAddressFromBytesImpl :: MaybeFfiHelper -> Uint8Array -> Maybe BaseAddressCsl
+foreign import baseAddressFromBech32Impl :: MaybeFfiHelper -> Bech32String -> Maybe BaseAddressCsl
+
+foreign import rewardAddressFromBytesImpl :: MaybeFfiHelper -> Uint8Array -> Maybe RewardAddressCsl
+foreign import rewardAddressFromBech32Impl :: MaybeFfiHelper -> Bech32String -> Maybe RewardAddressCsl
+
+-- | The exported constructor for `BaseAddress`.
+baseAddress
+  :: forall p d
+   . (ToCsl (StakeCred d) StakeCredentialCsl)
+  => (ToCsl (StakeCred p) StakeCredentialCsl)
+  => {network :: NetworkTag, payment :: p, delegation :: d}
+  -> BaseAddress p d
+baseAddress {network, payment, delegation} = BaseAddress
+  { network: network
+  , payment: payment
+  , delegation: delegation
+  }
+
+-- | The exported constructor for `RewardAddress`.
+rewardAddress
+  :: forall p
+   . (ToCsl (StakeCred p) StakeCredentialCsl)
+  => {network :: NetworkTag, payment :: p}
+  -> RewardAddress p
+rewardAddress {network, payment} = RewardAddress
+  { network: network
+  , payment: payment
+  , delegation: unit
+  }
+
+
+-- | Convert address to its byte representation
+addressBytes
+  :: forall a p d
+   . Address a p d
+  => a
+  -> Uint8Array
```

Comment:
I suggest using recently merged `ByteArray` everywhere, it has the same runtime representation as Uint8Array

Reply:
Okay